cmd/snap-failure/cmd_snapd.go

```diff
@@ -141,6 +147,23 @@ func (c *cmdSnapd) Execute(args []string) error {
 		return fmt.Errorf("snapd failed: %v", err)
 	}
 
+	isFailedCmd := runCmd("systemctl", []string{"is-failed", "snapd.socket", "snapd.service"}, nil)
```

Comment:
I think we should really make clear to ourselves whose responsibility it is to fix the situation, at this point I really fear that the source of these many bugs is that we have race conditions between snap-failure and the ephemeral snapd (and then the real snapd) trying to fix things. One way that we could do this is like you said (and maybe you do in the followup I haven't gotten that far yet) and really force the ephemeral snapd to not do too much except rewrite the systemd unit files and then exit, and let snap-failure take responsibility for restarting the real snapd. Or, we could take this snippet of code here and make this in a loop and have snap-failure take responsibility for restart snapd.service.

I don't have strong opinions on it either way, but I think this question of who should do what is definitely an underlying ulcer we should resolve before trying to make things more complicated on either side.

Reply:
I agree with he goal, but notice that this PR is very deliberate in not touching anything in snapd itself. The reality is that we have shipped snapds and because of the vagaries of sequence we might end up we a newer snap-failure calling an already shipped snapd, so we cannot just change this and snapd assuming they are somehow in sync. That's why the code here try to be both backward (tries to fix snap-failure with the snapd we have) and forward compatible.

Reply:
Yes, understood. Not for this PR or perhaps even the followup PR, but since we have the sequence.json could we encode more information there about how smart the ephemeral snapd is so that in the best case where a newish snap-failure is reverting to a still newish snapd we can "do the best thing"? I think it would be best to have more efficient branches in snap-failure logic so that over time in the real world when something fails the most robust logic path is used.

Reply:
I don't think we need to extend sequence for that,  we could use /usr/lib/snapd/info for that I suppose if needed. But yes is definitely an cleanup/optimisation for later at this point